design/src/rfcs/rfc0024_better_constraint_violations.md

```diff
@@ -0,0 +1,509 @@
+   constraint traits, setting its visibility to `pub (crate)`, for exclusive
+   use by the framework.
+2. [`ServerBuilderGeneratorWithoutPublicConstrainedTypes.kt`] renders the
+   builder type the user is exposed to: this builder does not take in
+   constrained types and does not enforce all modeled constraints.
+
+[`ServerBuilderGenerator.kt`]: https://github.com/awslabs/smithy-rs/blob/2226feff8f7fa884204f81a50d7e016386912acc/codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerBuilderGenerator.kt
+[`ServerBuilderGeneratorWithoutPublicConstrainedTypes.kt`]: https://github.com/awslabs/smithy-rs/blob/2226feff8f7fa884204f81a50d7e016386912acc/codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerBuilderGeneratorWithoutPublicConstrainedTypes.kt
+
+Collecting constraint violations
+--------------------------------
+
+### Background
+
+Constrained operations are currently required to have
+`smithy.framework#ValidationException` as a member in their [`errors`
+property](https://smithy.io/2.0/spec/service-types.html#operation). This is the
+shape that is rendered in responses when a request contains data that violates
+the modeled constraints.
+
+The shape is defined in the
+[`smithy-validation-model`](https://search.maven.org/artifact/software.amazon.smithy/smithy-validation-model)
+Maven package, [as
+follows](https://github.com/awslabs/smithy/blob/main/smithy-validation-model/model/smithy.framework.validation.smithy):
+
+```smithy
+$version: "2.0"
+
+namespace smithy.framework
+
+/// A standard error for input validation failures.
+/// This should be thrown by services when a member of the input structure
+/// falls outside of the modeled or documented constraints.
+@error("client")
+structure ValidationException {
+
+    /// A summary of the validation failure.
+    @required
+    message: String,
+
+    /// A list of specific failures encountered while validating the input.
+    /// A member can appear in this list more than once if it failed to satisfy multiple constraints.
+    fieldList: ValidationExceptionFieldList
+}
+
+/// Describes one specific validation failure for an input member.
+structure ValidationExceptionField {
+    /// A JSONPointer expression to the structure member whose value failed to satisfy the modeled constraints.
+    @required
+    path: String,
+
+    /// A detailed description of the validation failure.
+    @required
+    message: String
+}
+
+list ValidationExceptionFieldList {
+    member: ValidationExceptionField
+}
+```
+
+It was mentioned in the [constraint traits
+RFC](https://github.com/awslabs/smithy-rs/pull/1199#discussion_r809300673), and
+implicit in the definition of Smithy's
+[`smithy.framework.ValidationException`](https://github.com/awslabs/smithy/blob/main/smithy-validation-model/model/smithy.framework.validation.smithy)
+shape, that server frameworks should respond with a _complete_ collection of
```

Comment:
I'm not sure I agree with this interpretation.
I think it's fair to say that the framework should return an error that is as informative as possible, but it doesn't necessarily have to be complete.

It's bit like compiler errors, to use an analogy - you don't want a compiler to short-circuit as soon as it finds the first error, otherwise you need to run through the fix-check loop several times. At the same time, you value fast feedback and you'd prefer the compiler to return in a reasonable timeframe with a set of relevant errors rather than taking forever to return a complete list.

Going back to Smithy, I think it would be sensible to put in place some upper bounds (e.g. 1000 elements) to prevent degenerate behaviour (or DOS attacks, as mentioned below) and guarantee a timely response.

Reply:
The RFC now subscribes this opinion after the rewrite: https://github.com/awslabs/smithy-rs/pull/2040/commits/5033c7af05e7e1aa9fe2308cf52dd27fe0c3deee.